### ISSUE TITLE
Missing RBAC roles for pach service accounts

### ISSUE THREAD

**Issue by williballenthin** [pachyderm/pachyderm#2601]
During deploy, pachyderm does not ensure that appropriate roles are bound to its service accounts. This prevents pachyderm from working on k8s clusters with restrictive default permissions.

# Environment

  - AWS
  - k8s deployed via [kube-aws](https://github.com/kubernetes-incubator/kube-aws) according to the [manual](https://kubernetes-incubator.github.io/kube-aws/getting-started/step-1-configure.html)
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:28:34Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.4+coreos.0", GitCommit:"4292f9682595afddbb4f8b1483673449c74f9619", GitTreeState:"clean", BuildDate:"2017-11-21T17:22:25Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```
  - pachyderm 1.6.5
```
$ pachctl version
pachctl             1.6.5
pachd               1.6.5
```

# Behavior
After deploying k8s to AWS using the kube-aws tool and deploying pachyderm, the output from `kubectl get all` looked good. Creating jobs failed, however, with the error:

```
Reason: failed to create workers: replicationcontrollers is forbidden: User "system:serviceaccount:default:pachyderm" cannot create replicationcontrollers in the namespace "default"
```

This appears to be a problem the k8s RBAC (and not with AWS permissions).

# Reproducing

Deploy k8s to AWS using the kube-aws tool:

```
$ aws kms create-key
    --region=us-east-1 
    --description="kube-aws assets"
$ aws s3api create-bucket 
    --region=us-east-1 
    --bucket flare-...-v7
$ kube-aws init 
    --cluster-name=flare-...-v7 
    --external-dns-name="v7....com" 
    --hosted-zone-id="..." 
    --region=us-east-1 
    --availability-zone=us-east-1a 
    --key-name=... 
    --kms-key-arn="arn:aws:kms:us-east-1:..."
$ kube-aws render credentials 
    --generate-ca
$ kube-aws render stack
$ kube-aws validate 
    --s3-uri s3://flare-...-v7/
$ kube-aws up 
    --s3-uri s3://flare-...-v7/
```

and deploy pachyderm:

```
$ pachctl deploy amazon flare-...-v7 us-east-1 10 
    --dynamic-etcd-nodes=1 
    --iam-role flare-...-v7-Nodepool1-...VRETX
```

then create a simple pipeline, note that it fails using `pachctl list-pipeline`, and see why using `pachctl inspect-pipeline`. You should see something like:
```
Reason: failed to create workers: replicationcontrollers is forbidden: User "system:serviceaccount:default:pachyderm" cannot create replicationcontrollers in the namespace "default"
```

# Notes

## potential fix
I was able to fix the pipeline creation issue by creating a k8s rolebinding from the `default:pachyderm` service account to the `edit` role:

```
$ kubectl create rolebinding pach-edit --clusterrole=edit --serviceaccount=default:pachyderm --namespace=default
```

Perhaps during pachyderm deploy, the tool can query service account roles and add these bindings as needed.

## work in progress
My pipelines are not yet processing jobs, so there may be a few more permissions tweaks necessary. I will update here as I gather more information.

**Comment by williballenthin** [comment]
There's also this error in the k8s logs:

```
$ kubectl logs -l app=pachd | head -n1
time="2017-12-28T16:17:13Z" level=error msg="unable to access kubernetes nodeslist, Pachyderm will continue to work but it will not be possible to use COEFFICIENT parallelism. error: nodes is forbidden: User "system:serviceaccoun
t:default:pachyderm" cannot list nodes at the cluster scope"
```

**Comment by williballenthin** [comment]
I've reviewed the kubectl logs for `app=pachd` and `suite=pachyderm` and there are no other warning/error messages.

However, jobs are not spawning as expected (this workflow has worked on minikube and gce before), so i wonder if there are some other errors that have been swallowed. I'm not sure where else to look right now.

**Comment by JoeyZwicker** [comment]
Hi @williballenthin. Thanks for all the details in this issue. Now that everyone is getting from break, we'll take a look at this and figure out the best way forward for you.

**Comment by jdoliner** [comment]
@williballenthin thanks for the detailed bug report. I think the potential fix you propose is spot on so that's what we're going to do, this should be available in the 1.7 release.

**Comment by williballenthin** [comment]
Hey @JoeyZwicker and @jdoliner thanks for the update. Glad to hear that the potential fixes were on the right track.

While there are no other warning/error messages in the logs, and I can upload and list files to repos, jobs are not being spawned. But I don't know where to look to triage and confirm it is/n't related to the RBAC settings. Have you been able to reproduce this? Is there any information I can collect to chase it down? 

(We can migrate to a separate issue if you'd like)

**Comment by jdoliner** [comment]
@williballenthin what kubernetes state do you see when you create a pipeline? There should be a new `rc` created for the workers for your pipeline which will cause the pods to spin up. That's the thing that creates jobs so if you're not seeing it, there's still some sort of k8s problem. Otherwise it's a pachyderm bug.

**Comment by williballenthin** [comment]
Here's the first pipeline in my flow. It hashes and arranges input samples:

```
$ pachctl list-pipeline
NAME           INPUT              OUTPUT                CREATED             STATE
sample         samples:/*         sample/master         7 days ago          running

$ pachctl inspect-pipeline sample
Name: sample
Created: 7 days ago
State: running
Reason:
Parallelism Spec: <nil>
ResourceRequests:
        CPU: 0
        Memory: 64M

Input:
...

...
Job Counts:
starting: 0     running: 0      failure: 0      success: 0
```

Note the job count is 0, despite the fact I've added files to the input repository:

```
$ pachctl list-job
ID OUTPUT COMMIT STARTED DURATION RESTART PROGRESS DL UL STATE
[empty]

$ pachctl list-file samples master
NAME                TYPE                SIZE
cat                 file                50.86KiB
ls                  file                123.6KiB
```

Here is the k8s state. It seems there is a `rc` for the pipeline.

```
$ kubectl get all
NAME                  DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deploy/pachd          1         1         1            1           7d

NAME                         DESIRED   CURRENT   READY     AGE
rs/pachd-64cb8cffb8          1         1         1         7d

NAME                  DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deploy/pachd          1         1         1            1           7d

NAME                DESIRED   CURRENT   AGE
statefulsets/etcd   1         1         7d

NAME                         DESIRED   CURRENT   READY     AGE
rs/pachd-64cb8cffb8          1         1         1         7d

NAME                                        READY     STATUS    RESTARTS   AGE
po/etcd-0                                   1/1       Running   0          7d
po/pachd-64cb8cffb8-mrq85                   1/1       Running   0          1d
po/pipeline-sample-v1-8xs66                 2/2       Running   0          7d

NAME                                  DESIRED   CURRENT   READY     AGE
rc/pipeline-sample-v1                 1         1         1         7d

NAME                                   TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                                                   AGE
svc/etcd                               NodePort    10.3.0.179   <none>        2379:32096/TCP                                            7d
svc/etcd-headless                      ClusterIP   None         <none>        2380/TCP                                                  7d
svc/kubernetes                         ClusterIP   10.3.0.1     <none>        443/TCP                                                   8d
svc/pachd                              NodePort    10.3.0.42    <none>        650:30650/TCP,651:30651/TCP,652:30652/TCP,999:30999/TCP   7d
svc/pipeline-sample-v1                 ClusterIP   10.3.0.239   <none>        80/TCP                                                    7d
```

**Comment by williballenthin** [comment]
fetching logs from the worker pod shows lots of errors (its been running for a week):

```
$ kubectl logs pipeline-sample-v1-8xs66 user | head
{"pipelineName":"sample","workerId":"pipeline-sample-v1-8xs66","master":true,"ts":"2017-12-28T17:32:02.692277370Z","message":"Launching worker master process"}
{"pipelineName":"sample","workerId":"pipeline-sample-v1-8xs66","master":true,"ts":"2017-12-28T17:32:02.704355432Z","message":"master: error running the master process: master: error constructing worker pool: unknown (get endpoints); retrying in %!!(MISSING)v(MISSING); retrying in 8.608714168s"}
{"pipelineName":"sample","workerId":"pipeline-sample-v1-8xs66","master":true,"ts":"2017-12-28T17:32:11.316102829Z","message":"Launching worker master process"}
{"pipelineName":"sample","workerId":"pipeline-sample-v1-8xs66","master":true,"ts":"2017-12-28T17:32:11.320513340Z","message":"master: error running the master process: master: error constructing worker pool: unknown (get endpoints); retrying in %!!(MISSING)v(MISSING); retrying in 16.060099141s"}
[...and 93,000 similar lines...]
```

**Comment by JoeyZwicker** [comment]
@williballenthin We have a PR in progress that should fix this issue and get Pachyderm to work with RBAC better.

**Comment by jdoliner** [comment]
Fixed in #2605.
